xcube/core/store/preload.py

```diff
@@ -0,0 +1,409 @@
+# Copyright (c) 2018-2024 by xcube team and contributors
+# Permissions are hereby granted under the terms of the MIT License:
+# https://opensource.org/licenses/MIT.
+
+from abc import abstractmethod, ABC
+from asyncio import CancelledError
+from concurrent.futures import Executor, Future
+from concurrent.futures.thread import ThreadPoolExecutor
+from enum import Enum
+import threading
+from typing import Any, Callable
+
+import tabulate
+
+from xcube.util.assertions import assert_given, assert_instance
+
+
+class PreloadStatus(Enum):
+    """Preload process status."""
+
+    waiting = "waiting"
+    started = "started"
+    stopped = "stopped"
+    cancelled = "cancelled"
+    failed = "failed"
+
+    def __str__(self):
+        return self.name.upper()
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}.{self.name}"
+
+
+class PreloadState:
+    """Preload state."""
+
+    def __init__(
+        self,
+        data_id: str,
+        status: PreloadStatus | None = None,
+        progress: float | None = None,
+        message: str | None = None,
+        exception: BaseException | None = None,
+    ):
+        assert_given(data_id, name="data_id")
+        self.data_id = data_id
+        self.status = status
+        self.progress = progress
+        self.message = message
+        self.exception = exception
+
+    def update(self, event: "PreloadState"):
+        """Update this state with the given partial state.
+
+        Args:
+            event: the partial state.
+        """
+        assert_instance(event, PreloadState, name="data_id")
+        if self.data_id == event.data_id:
+            if event.status is not None:
+                self.status = event.status
+            if event.progress is not None:
+                self.progress = event.progress
+            if event.message is not None:
+                self.message = event.message
+            if event.exception is not None:
+                self.exception = event.exception
+
+    def __str__(self):
+        return ", ".join(f"{k}={v}" for k, v in _to_dict(self).items())
+
+    def __repr__(self):
+        args = ", ".join(f"{k}={v!r}" for k, v in _to_dict(self).items())
+        return f"{self.__class__.__name__}({args})"
+
+
+class PreloadHandle(ABC):
```

Comment:
We also talked about the `cache_data_store` and the `subscribe` method. Will this be added in the next PRs?

Reply:
- `cache_data_store` is an implementation detail
- `subscribe` can be added if needed. The current design and programming model